### PR TITLE
fix(forge): format `new` expressions

### PR DESCRIPTION
**Pull request by beeb**
The formatter currently doesn't handle `new` expressions, which leads to https://github.com/foundry-rs/foundry/issues/4619.

This PR fixes that by adding the relevant case in the formatter. The tests have been adjusted to reflect this. They do not pass without the fix, but pass with the fix added.

Please note that the existing test cases had `new uint[](length)` in several places, where the type is not normalized to `uint256`. In my opinion, the behavior after the fix (convert those to `uint256`) is correct, and the test files were wrong.

Thus, I adjusted all occurences of `new uint[]` in the `fmt.sol` files.

A new test case for `new Contract()` expressions has been added to the `FunctionCall` test, since those expression can be seen as a function call to the constructor.

NOTE: cargo fmt changed some formatting in `formatter.rs`, no sure if that's ok.

## Motivation

The formatter is broken for expression which start with `new`.

## Solution

Handle the case where an expression is `Expression::New`.

Closes https://github.com/foundry-rs/foundry/issues/4619

### CODE REVIEW

File: crates/fmt/src/formatter.rs [hunk 1]
```diff
@@ -2286,6 +2287,10 @@ impl<'a, W: Write> Visitor for Formatter<'a, W> {
                 expr.visit(self)?;
                 stmt.visit(self)?;
             }
+            Expression::New(_, expr) => {
+                write!(self.buf(), "new ")?;
+                self.visit_expr(expr.loc(), expr)?;
+            }
```
Comment (beeb) [inline crates/fmt/src/formatter.rs:2293]: Here is the fix

Comment (mattsse) [reply crates/fmt/src/formatter.rs:2292]: oh nice!

**Review by mattsse — APPROVED** [review]
ty!

lgtm

**Review by mattsse — CHANGES_REQUESTED** [review]
this has a smol regression, see failing tests:

```
ConstructorVictim victim =new ConstructorVictim(
```
There's a whitespace missing that I think we can check for

I believe before, new expressions where formatted as is, but I like visiting the expression better

**Comment by beeb** [comment]
The failing workflow indicates that there is potentially still a problem with the implementation. But the last two reported mismatches (in `testdata/cheats/RecordAccountAccesses.t.sol`) are IMO badly formatted in main currently.

**Comment by mattsse** [comment]
> But the last two reported mismatches (in testdata/cheats/RecordAccountAccesses.t.sol) are IMO badly formatted in main currently.

agree because previously it formatted using the src directly

**Comment by beeb** [comment]
@mattsse I had a deeper look and the problem only happens when the right side of the assignment is a  `Expression::New` and doesn't fit on a single line.
Do you have an idea why there is no space in that case? I can't figure it out rn.

**Comment by mattsse** [comment]
I might have, I'll have a look.

**Comment by beeb** [comment]
Ok got it!

**Comment by mattsse** [comment]
amazing, ty!

**Comment by beeb** [comment]
Regarding formatting, do you know why my nightly rustfmt added semicolons for the return statements?

**Comment by beeb** [comment]
Ah I figured it out. Actually I formatted wrongly with rustfmt stable at first, which added semicolons, and then switched back to nightly and it kept the semicolons. But if they are not there to begin with it doesn't add them.

**Comment by mattsse** [comment]
the recent nightly fmt appears to be broken to some extent at least on macos

I'm currently using `+nightly-2023-11-14` which is not affected but also does not respect trailing_semicolon...

I noticed this 2 weeks ago or so, let's hope gets fixed soon

**Pull request merged**